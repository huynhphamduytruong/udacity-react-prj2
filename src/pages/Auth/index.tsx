import LoginLogo from 'images/LoginLogo'
import { Card, Row, Col } from 'react-bootstrap'
import { Outlet } from 'react-router-dom'
import { StyledAuthPage, StyledAuthContainer } from './Auth.styled'

export const AuthPage = () => {
  return (
    <StyledAuthPage>
      <StyledAuthContainer>
        <Card body className="d-grid shadow-lg p-3 rounded h-100">
          <Row className="justify-content-center align-items-center h-100">
            <Col lg={6} className="d-flex justify-content-center align-items-center">
              <LoginLogo />
            </Col>
            <Col lg={6}>
              <div className="px-5 ms-xl-4">
                <Outlet />
              </div>
            </Col>
          </Row>
        </Card>

        <div className="text-center footer mt-4">
          Coded with <p className="text-danger d-inline-block beating">&hearts;</p> by{' '}
          <a href="https://github.com/huynhphamduytruong/udacity-react-prj2">TruongHPD</a>.
        </div>
      </StyledAuthContainer>
    </StyledAuthPage>
  )
}
