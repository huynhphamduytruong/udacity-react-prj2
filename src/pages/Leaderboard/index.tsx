import { useAppDispatch, useAppSelector } from 'app/hooks'
import { getUsers, usersSelector } from 'features/User/userSlice'
import { useEffect } from 'react'
import { Spinner, Table } from 'react-bootstrap'
import { User } from 'types'

export const Leaderboard = () => {
  const usersStore = useAppSelector(usersSelector)

  const { loading: usersLoading, users } = usersStore

  const dispatch = useAppDispatch()

  useEffect(() => {
    void dispatch(getUsers())
  }, [])

  // Sum both the number of questions and answers and use the sum to compare with other
  const sortLeaderboard = (u1: User, u2: User) =>
    Object.keys(u2.answers).length + u2.questions.length - Object.keys(u1.answers).length - u1.questions.length

  if (usersLoading) return <Spinner />

  return (
    <Table striped>
      <thead>
        <tr>
          <th>#</th>
          <th>User</th>
          <th>Answered</th>
          <th>Created</th>
        </tr>
      </thead>
      <tbody>
        {users
          ?.slice()
          .sort(sortLeaderboard)
          .map((u, idx) => (
            <tr key={idx}>
              <td>{++idx}</td>
              <td>
                <div>
                  {u.name} <span className="text-muted">({u.id})</span>
                </div>
              </td>
              <td>{Object.keys(u.answers).length}</td>
              <td>{u.questions.length}</td>
            </tr>
          ))}
      </tbody>
    </Table>
  )
}
