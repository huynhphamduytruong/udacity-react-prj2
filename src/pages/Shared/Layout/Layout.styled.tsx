import styled from 'styled-components'

export const StyledLayout = styled.div.attrs(({ className }) => ({
  className: `vh-100 ${className ?? ''}`.trim()
}))`
  main > .container {
    padding: 75px 0;
  }
`
