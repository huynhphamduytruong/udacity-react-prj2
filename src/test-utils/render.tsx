import React from 'react'
import { render as RTLRender, RenderOptions } from '@testing-library/react'
import { GlobalStyle } from 'index.styled'
import { Provider } from 'react-redux'
import store, { AppStore, RootState, setupStore } from 'app/store'
import { MemoryRouter, RouterProvider, createMemoryRouter } from 'react-router-dom'
import type { Router } from '@remix-run/router'
import { PropsWithChildren } from 'react'
import { PreloadedState } from '@reduxjs/toolkit'
import { AppRoutes } from 'app/routes'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  router?: Router
}

export function renderWithProvider(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    router = createMemoryRouter(AppRoutes),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <MemoryRouter>
        <Provider store={store}>{children}</Provider>
      </MemoryRouter>
    )
  }
  return { store, ...RTLRender(ui, { wrapper: Wrapper, ...renderOptions }) }
}

const TestAppWithReduxAndRouter = ({ store, router }: { store?: AppStore; router?: Router }) => {
  return (
    <Provider store={store}>
      <GlobalStyle />
      <RouterProvider router={router} />
    </Provider>
  )
}

export const renderWithRouter = (router, options?: Parameters<typeof RTLRender>[1]) => {
  const rendered = RTLRender(<TestAppWithReduxAndRouter router={router} store={store} />, options)
  return {
    ...rendered,
    rerender: (ui, options?: Parameters<typeof RTLRender>[1]) =>
      renderWithRouter(router, { container: rendered.container, ...options })
  }
}
