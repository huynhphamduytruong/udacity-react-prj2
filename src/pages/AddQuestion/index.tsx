import { Container, Button, Form, FloatingLabel, Spinner, Toast, ToastContainer } from 'react-bootstrap'
import { useForm } from 'react-hook-form'
import { authSelector } from 'features/Auth/authSlice'
import { useAppDispatch, useAppSelector } from 'app/hooks'
import { saveQuestion } from 'features/Poll/pollSlice'
import { useState } from 'react'
import { BiCommentAdd } from 'react-icons/bi'

interface NewQuestionData {
  option1: string
  option2: string
}

export const AddQuestion = () => {
  const authStore = useAppSelector(authSelector)

  const { user } = authStore
  const dispatch = useAppDispatch()

  const [loading, setLoading] = useState(false)
  const [showToast, setShowToast] = useState(false)

  const {
    register,
    handleSubmit,
    formState: { errors, defaultValues, isValid },
    resetField
  } = useForm<NewQuestionData>({
    defaultValues: {
      option1: null,
      option2: null
    }
  })

  const onSubmit = (data) => {
    setLoading(true)
    setShowToast(false)

    void dispatch(
      saveQuestion({ author: user.id, options: { optionOne: data.option1, optionTwo: data.option2 } })
    ).then(() => {
      setLoading(false)
      setShowToast(true)
      resetField('option1')
      resetField('option2')
    })
  }

  return (
    <Container className="my-5">
      <ToastContainer className="p-3" position={'top-end'} style={{ zIndex: 999, marginTop: '70px' }}>
        <Toast onClose={() => setShowToast(false)} show={showToast} delay={3000} autohide>
          <Toast.Header closeButton={false}>
            <BiCommentAdd className="me-2" />
            <strong className="me-auto">New question</strong>
          </Toast.Header>
          <Toast.Body>Add question successfully</Toast.Body>
        </Toast>
      </ToastContainer>

      <h1 className="text-center mb-4">Would You Rather?</h1>

      <Form className="custom-form" onSubmit={handleSubmit(onSubmit)}>
        <FloatingLabel controlId="username" label="Enter your first option" className="mb-3">
          <Form.Control
            data-testid="option1"
            name="option1"
            type="text"
            placeholder="Enter your first option"
            defaultValue={defaultValues.option1}
            aria-invalid={errors.option1 ? 'true' : 'false'}
            {...register('option1', { required: 'This field is required' })}
          />
          {errors.option1 && <Form.Text className="text-danger">{errors.option1.message}</Form.Text>}
        </FloatingLabel>

        <hr className="hr-text" data-content="OR" />

        <FloatingLabel controlId="username" label="Enter your second option" className="mb-3">
          <Form.Control
            data-testid="option2"
            name="option2"
            type="text"
            placeholder="Enter your second option"
            defaultValue={defaultValues.option2}
            aria-invalid={errors.option2 ? 'true' : 'false'}
            {...register('option2', { required: 'This field is required' })}
          />
          {errors.option2 && <Form.Text className="text-danger">{errors.option2.message}</Form.Text>}
        </FloatingLabel>

        <div className="text-center">
          <Button data-testid="submit-btn" variant="primary" type="submit" disabled={loading || !isValid}>
            {loading ? (
              <>
                <Spinner as="span" animation="border" size="sm" role="status" aria-hidden="true" />
                <span className="visually-hidden">Loading...</span>
              </>
            ) : (
              <>{'Submit'}</>
            )}
          </Button>
        </div>
      </Form>
    </Container>
  )
}
