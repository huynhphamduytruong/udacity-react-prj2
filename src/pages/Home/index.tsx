import { useEffect, useState, useMemo } from 'react'
import { Badge, Tabs, Tab, Col, Row, Spinner, Toast, ToastContainer } from 'react-bootstrap'
import { GiVote } from 'react-icons/gi'

import { QuestionContainer } from './Home.styled'
import { QuestionCard } from 'components/QuestionCard'
import { useAppDispatch, useAppSelector } from 'app/hooks'
import { getPolls, pollsSelector, saveAnswer } from 'features/Poll/pollSlice'
import { authSelector } from 'features/Auth/authSlice'
import { getUsers, usersSelector } from 'features/User/userSlice'

export const HomePage = () => {
  const usersStore = useAppSelector(usersSelector)
  const pollsStore = useAppSelector(pollsSelector)
  const authStore = useAppSelector(authSelector)
  const dispatch = useAppDispatch()

  const { loading: pollsLoading, polls } = pollsStore
  const { loading: usersLoading, users } = usersStore
  const { user } = authStore

  const [showToast, setShowToast] = useState(false)

  const ansPolls = useMemo(
    () => polls.filter((p) => Object.keys(users.find((u) => u.id === user.id)?.answers ?? {}).some((a) => a === p.id)),
    [polls, user.id, users]
  )
  const unansPolls = useMemo(() => polls.filter((p) => !ansPolls.includes(p)), [polls, ansPolls])

  useEffect(() => {
    void dispatch(getPolls())
    void dispatch(getUsers())
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const handleVoting = async (qid, option) => {
    void dispatch(saveAnswer({ author: user.id, questionId: qid, option: option })).then(() => {
      setShowToast(true)
    })
  }

  return (
    <>
      <ToastContainer className="p-3" position={'top-end'} style={{ zIndex: 999, marginTop: '70px' }}>
        <Toast onClose={() => setShowToast(false)} show={showToast} delay={3000} autohide>
          <Toast.Header closeButton={true}>
            <GiVote className="me-2" />
            <strong className="me-auto">Vote</strong>
          </Toast.Header>
          <Toast.Body>Vote successfully</Toast.Body>
        </Toast>
      </ToastContainer>

      {pollsLoading || usersLoading ? (
        <Spinner />
      ) : (
        <Tabs defaultActiveKey="unansw">
          <Tab
            eventKey="unansw"
            title={
              <span className="text-primary">
                Unanswered Questions <Badge bg="primary">{unansPolls.length}</Badge>
              </span>
            }
          >
            <QuestionContainer>
              <Row md={1} lg={2} className="g-3">
                {unansPolls.map((x, idx) => (
                  <Col md={12} lg={6} key={idx}>
                    <QuestionCard
                      question={x}
                      questioner={users.find((u) => u.id === x.author)}
                      enabled
                      onVote={handleVoting}
                    />
                  </Col>
                ))}
              </Row>
            </QuestionContainer>
          </Tab>
          <Tab
            eventKey="answ"
            title={
              <span className="text-success">
                Answered Questions <Badge bg="success">{ansPolls.length}</Badge>
              </span>
            }
          >
            <QuestionContainer>
              <Row md={1} lg={2} className="g-3">
                {ansPolls.map((x, idx) => (
                  <Col md={12} lg={6} key={idx}>
                    <QuestionCard
                      question={x}
                      questioner={users.find((u) => u.id === x.author)}
                      selectedOption={users.find((u) => u.id === user.id)?.answers?.[x.id]}
                    />
                  </Col>
                ))}
              </Row>
            </QuestionContainer>
          </Tab>
        </Tabs>
      )}
    </>
  )
}
