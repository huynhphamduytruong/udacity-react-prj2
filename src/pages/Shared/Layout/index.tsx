import { Container } from 'react-bootstrap'
import { StyledLayout } from './Layout.styled'

import AppMenu from 'components/AppMenu'
import { AppFooter } from 'components/AppFooter'
import { PropsWithChildren } from 'react'

export type ILayoutProps = PropsWithChildren<unknown>

const Layout = ({ children }: ILayoutProps) => {
  return (
    <StyledLayout className="d-flex flex-column">
      <AppMenu />
      <main className="flex-shrink-0">
        <Container>{children}</Container>
      </main>
      <AppFooter />
    </StyledLayout>
  )
}

export const PublicLayout = ({ children }: ILayoutProps) => {
  return <StyledLayout>{children}</StyledLayout>
}

export default Layout
