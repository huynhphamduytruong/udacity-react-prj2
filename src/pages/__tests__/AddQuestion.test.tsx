import '@testing-library/jest-dom/extend-expect'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { AddQuestion } from 'pages/AddQuestion'
import Layout from 'pages/Shared/Layout'
import { renderWithProvider } from 'test-utils/render'

describe('AddQuestion page', () => {
  const preloadedState = {
    auth: { isAuthenticated: true, user: { id: ':id:', name: ':name:', avatarURL: ':avatarURL:' } }
  }

  test('should match snapshot', () => {
    const { container } = renderWithProvider(
      <Layout>
        <AddQuestion />
      </Layout>,
      { preloadedState }
    )

    expect(container).toMatchSnapshot()
  })

  test('should enable the submit button when both inputs have value', async () => {
    renderWithProvider(
      <Layout>
        <AddQuestion />
      </Layout>,
      { preloadedState }
    )

    const user = userEvent.setup()

    const inputOne = screen.getByTestId('option1')
    const inputTwo = screen.getByTestId('option2')
    const submitButton = screen.getByTestId('submit-btn')

    expect(submitButton).toHaveAttribute('disabled')

    await user.type(inputOne, 'first value')
    await user.type(inputTwo, 'second value')

    expect(submitButton).not.toHaveAttribute('disabled')
  })
})
