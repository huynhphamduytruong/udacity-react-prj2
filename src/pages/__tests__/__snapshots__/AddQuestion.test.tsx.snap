// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddQuestion page should match snapshot 1`] = `
<div>
  <div
    class="sc-bcPKhP cRFdXx vh-100 d-flex flex-column"
  >
    <header
      class="sc-grXZZQ lcmHQc fixed-top bg-body-tertiary mb-2"
    >
      <nav
        class="navbar navbar-expand-md navbar-light"
      >
        <div
          class="container"
        >
          <a
            class="home-link me-3 navbar-brand"
            href="/"
          >
            <svg
              height="0"
              width="0"
            >
              <lineargradient
                id="blue-gradient"
                x1="100%"
                x2="0%"
                y1="100%"
                y2="0%"
              >
                <stop
                  offset="0%"
                  stop-color="rgba(246, 128, 132, 0.9)"
                />
                <stop
                  offset="100%"
                  stop-color="rgba(166, 192, 254, 0.9)"
                />
              </lineargradient>
            </svg>
            <svg
              class="me-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="fill: url(#blue-gradient); font-size: 1.6rem;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M494 21.621c-14.947 8.43-29.566 17.581-43.67 29.227l7.318 38.547C471.923 93.66 483.583 95.26 494 95.36v-73.74zm-98.982 24.512c-15.283-.085-32.48 2.596-53.832 6.834l-.22.043-.22.033c-14.77 2.177-40.794 12.065-66.465 38.867l44.27 11.766c.972-1.493 5.936-9.004 6.88-10.555 5.124 3.123 10.248 6.244 15.372 9.365-12.475 20.475-26.742 35.556-43.934 54.522-2.123 4.718.977 8.199 4.36 10.14 5.22 2.931 14.1 3.09 16.437 2.102 23.932-15.768 40.819-35.928 55.963-56.271l5.469.964c11.501 2.031 26.47 1.058 38.707-2.853 11.098-3.548 19.272-9.357 22.662-15.688L432.54 53.65c-12.044-5.214-24.039-7.442-37.523-7.517zM227.932 98.717l-29.436 115.986 9.643.297H311.27l.9-.209 6.804-27.092c-8.86 1.9-18.296-.217-26.557-4.855-5.188-2.913-10.024-7.24-12.621-13.434-7.797-19.938 15.857-37.297 28.724-52.75l-80.59-17.943zM69.562 201l-23 46h418.875l-23-46H334.195l-3.517 14H352v18H160v-18h19.852l3.552-14H69.563zM41 265v222h430V265H41zm14 14h402v194H55V279zm18 18v118.238l34.502-74.994 73.36 31.762 66.652-45.84 37.513 57.273 50.11-4.595 31.3-39.332L439 394.627V297H73zm169.543 54.43l-90.63 62.33 27.01 41.24h95.606l19.666-24.71-51.652-78.86zm-126.045 12.326L74.521 455h82.885l-30.193-46.098 36.144-24.857-46.859-20.29zm253.065.732L297.533 455h140.54l.927-1.166v-36.602l-69.438-52.744zm-49.944 33.854l-23.426 2.148 9.805 14.969 13.621-17.117z"
              />
            </svg>
            Employee Polls
          </a>
          <button
            aria-controls="nav-menu"
            aria-label="Toggle navigation"
            class="justify-content-end navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="nav-menu"
          >
            <div
              class="me-auto navbar-nav"
            >
              <a
                aria-current="page"
                class="nav-link active"
                data-rr-ui-event-key="/"
                href="/"
              >
                Home
              </a>
              <a
                class="nav-link"
                data-rr-ui-event-key="/leaderboard"
                href="/leaderboard"
              >
                Leaderboard
              </a>
              <a
                class="nav-link"
                data-rr-ui-event-key="/add"
                href="/add"
              >
                New Question
              </a>
            </div>
          </div>
          <div
            class="justify-content-end navbar-collapse collapse"
          >
            <div
              class="nav-item dropdown"
            >
              <a
                aria-expanded="false"
                class="dropdown-toggle nav-link"
                href="#"
                id="user-actions"
                role="button"
                tabindex="0"
              >
                <button
                  class="btn btn-none"
                  type="button"
                >
                  <img
                    alt=":id:"
                    height="36"
                    src="https://api.dicebear.com/6.x/big-smile/svg?seed=:id:&size=36"
                    width="36"
                  />
                   
                  :name:
                </button>
              </a>
            </div>
          </div>
        </div>
      </nav>
    </header>
    <main
      class="flex-shrink-0"
    >
      <div
        class="container"
      >
        <div
          class="my-5 container"
        >
          <div
            class="toast-container top-0 end-0 p-3"
            style="z-index: 999; margin-top: 70px;"
          />
          <h1
            class="text-center mb-4"
          >
            Would You Rather?
          </h1>
          <form
            class="custom-form"
          >
            <div
              class="mb-3 form-floating"
            >
              <input
                aria-invalid="false"
                class="form-control"
                data-testid="option1"
                id="username"
                name="option1"
                placeholder="Enter your first option"
                type="text"
                value=""
              />
              <label
                for="username"
              >
                Enter your first option
              </label>
            </div>
            <hr
              class="hr-text"
              data-content="OR"
            />
            <div
              class="mb-3 form-floating"
            >
              <input
                aria-invalid="false"
                class="form-control"
                data-testid="option2"
                id="username"
                name="option2"
                placeholder="Enter your second option"
                type="text"
                value=""
              />
              <label
                for="username"
              >
                Enter your second option
              </label>
            </div>
            <div
              class="text-center"
            >
              <button
                class="btn btn-primary"
                data-testid="submit-btn"
                disabled=""
                type="submit"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </main>
    <footer
      class="footer mt-auto py-3 bg-light fixed-bottom"
    >
      <div
        class="container"
      >
        <span
          class="text-muted"
        >
          Udacity React Nanodegree - Project 2 - 
          2023
           Â© TruongHPD
        </span>
      </div>
    </footer>
  </div>
</div>
`;
